packages/cw721/src/execute.rs

```diff
@@ -0,0 +1,810 @@
+                token_id,
+                expires,
+            } => self.approve(deps, env, info, spender, token_id, expires),
+            Cw721ExecuteMsg::Revoke { spender, token_id } => {
+                self.revoke(deps, env, info, spender, token_id)
+            }
+            Cw721ExecuteMsg::ApproveAll { operator, expires } => {
+                self.approve_all(deps, env, info, operator, expires)
+            }
+            Cw721ExecuteMsg::RevokeAll { operator } => self.revoke_all(deps, env, info, operator),
+            Cw721ExecuteMsg::TransferNft {
+                recipient,
+                token_id,
+            } => self.transfer_nft(deps, env, info, recipient, token_id),
+            Cw721ExecuteMsg::SendNft {
+                contract,
+                token_id,
+                msg,
+            } => self.send_nft(deps, env, info, contract, token_id, msg),
+            Cw721ExecuteMsg::Burn { token_id } => self.burn_nft(deps, env, info, token_id),
+            #[allow(deprecated)]
+            Cw721ExecuteMsg::UpdateOwnership(action) => {
+                self.update_minter_ownership(deps, env, info, action)
+            }
+            Cw721ExecuteMsg::UpdateMinterOwnership(action) => {
+                self.update_minter_ownership(deps, env, info, action)
+            }
+            Cw721ExecuteMsg::UpdateCreatorOwnership(action) => {
+                self.update_creator_ownership(deps, env, info, action)
+            }
+            Cw721ExecuteMsg::Extension { msg: _ } => Ok(Response::default()),
+            Cw721ExecuteMsg::SetWithdrawAddress { address } => {
+                self.set_withdraw_address(deps, &info.sender, address)
+            }
+            Cw721ExecuteMsg::RemoveWithdrawAddress {} => {
+                self.remove_withdraw_address(deps.storage, &info.sender)
+            }
+            Cw721ExecuteMsg::WithdrawFunds { amount } => self.withdraw_funds(deps.storage, &amount),
+        }
+    }
+
+    fn migrate(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        msg: Cw721MigrateMsg,
+        contract_name: &str,
+        contract_version: &str,
+    ) -> Result<Response, Cw721ContractError> {
+        let response = Response::<Empty>::default();
+        // first migrate legacy data ...
+        let response =
+            migrate_legacy_minter_and_creator(deps.storage, deps.api, &env, &msg, response)?;
+        let response = migrate_legacy_collection_info(deps.storage, &env, &msg, response)?;
+        let response = migrate_legacy_tokens(deps.storage, &env, &msg, response)?;
+        // ... then migrate
+        let response = migrate_version(deps.storage, contract_name, contract_version, response)?;
+        // ... and update creator and minter AFTER legacy migration
+        let response = migrate_creator(deps.storage, deps.api, &env, &msg, response)?;
+        let response = migrate_minter(deps.storage, deps.api, &env, &msg, response)?;
+        Ok(response)
+    }
+
```

Comment:
Here are the migration steps.

First do legacy migration:
1. migrate legacy minter, and set new creator with minter addr
2. migrate contract info to new collection info
3. migrate tokens to new nft store

Then normal migration:
4. set version
5. reset creator (if given)
6. reset minter (if given)

Please review especially legacy migrations thoroughly: ...

Reply:
@humanalgorithm look here for migration logic. It covers it all.